### ISSUE TITLE
Some calling routines do not properly handle SsINTERNAL, C_ERROR, and/or FALSE returned by low-level routines

### ISSUE THREAD

**Issue by mwinkel-dev** [MDSplus/mdsplus#2741]
**Affiliation**
MIT PSFC

**Version(s) Affected**
all

**Platform**
all

**Describe the bug**
The entire MDSplus code base uses different standards for error codes.
- on Linux, calls to the operating system functions usually return `-1` to indicate an error (i.e., access `errno` for the code)
- some MDSplus functions can return the macros `C_OK` (=0) and `C_ERROR` (= -1)
- others return `TRUE` (=1) or `FALSE`(=0)
- some return `B_TRUE` (=1) or `B_FALSE` (=0)
- and several return `SsINTERNAL` (= -1)
- plus many return MDSplus codes, such as `MDSplusSUCCESS` and `MDSplusERROR` (there are many more too)

Often the status returned by a function is immediately checked with one of the following "define" macros.

- `IS_OK`
- `STATUS_OK`
- `IS_NOT_OK`
- `STATUS_NOT_OK`

However, the above macros only work with MDSplus codes, which consist of a 32-bit integer containing 3 fields (facility ID, message ID, and severity code).  **Note that the low-order bit is important.  Success = 1 in the bit, Failure = 0 in the bit.**

The problems arise when non-MDSplus codes are passed to the macros.   Specifically, the -1 of `SsINTERNAL`, `C_ERROR` or an operating system error status.   **The value `-1` = `0xFFFFFFFF` has the low-order bit set, thus the macros will incorrectly treat `-1` as a flavor of success.** Which can cause error handling code to produce different results than what was  intended when the code was originally written.

This mix of error code conventions is a systemic source of errors in the entire code base.   It would probably be worthwhile to  at least make sure that all routines that return `SsINTERNAL` or `C_ERROR`, are in turn called by functions that properly detect those return values and then map them into the appropriate MDSplus error codes.

**To Reproduce**
This is an example that was spotted during the investigation of Issue #2731 and PR #2740.   The low-level routine, `send_bytes()` can return `SsINTERNAL`.   And potentially, it can ripple up the call stack all the way to the top level of `mdstcl`.  

The following links are to lines of code starting in `send_bytes()` and then climbing up the call stack.
[send_bytes()](https://github.com/MDSplus/mdsplus/blob/ad08c9cc7c88f7eed946bf06f2bd6a90e2e1a25f/mdstcpip/mdsipshr/SendMdsMsg.c#L71)
[SendMdsMsgC()](https://github.com/MDSplus/mdsplus/blob/ad08c9cc7c88f7eed946bf06f2bd6a90e2e1a25f/mdstcpip/mdsipshr/SendMdsMsg.c#L120)
[SendArg()](https://github.com/MDSplus/mdsplus/blob/ad08c9cc7c88f7eed946bf06f2bd6a90e2e1a25f/mdstcpip/mdsipshr/SendArg.c#L105)
[- bypasses error handling](https://github.com/MDSplus/mdsplus/blob/ad08c9cc7c88f7eed946bf06f2bd6a90e2e1a25f/mdstcpip/mdsipshr/SendArg.c#L107)
[GetAnswerInfoTS() can also return SsINTERNAL](https://github.com/MDSplus/mdsplus/blob/ad08c9cc7c88f7eed946bf06f2bd6a90e2e1a25f/servershr/ServerSendMessage.c#L203)
[- bypasses error handling](https://github.com/MDSplus/mdsplus/blob/ad08c9cc7c88f7eed946bf06f2bd6a90e2e1a25f/servershr/ServerSendMessage.c#L220)
[ServerSendMessage()](https://github.com/MDSplus/mdsplus/blob/ad08c9cc7c88f7eed946bf06f2bd6a90e2e1a25f/servershr/ServerSendMessage.c#L196)
[- bypasses error handling](https://github.com/MDSplus/mdsplus/blob/ad08c9cc7c88f7eed946bf06f2bd6a90e2e1a25f/servershr/ServerSendMessage.c#L197)
[ServerDispatchAction()](https://github.com/MDSplus/mdsplus/blob/ad08c9cc7c88f7eed946bf06f2bd6a90e2e1a25f/servershr/ServerDispatchAction.c#L67)
[ServerDispatchPhase()](https://github.com/MDSplus/mdsplus/blob/ad08c9cc7c88f7eed946bf06f2bd6a90e2e1a25f/servershr/ServerDispatchPhase.c#L638)
[- bypasses error handling](https://github.com/MDSplus/mdsplus/blob/ad08c9cc7c88f7eed946bf06f2bd6a90e2e1a25f/servershr/ServerDispatchPhase.c#L642)
[TclDispatch_phase()](https://github.com/MDSplus/mdsplus/blob/ad08c9cc7c88f7eed946bf06f2bd6a90e2e1a25f/tcl/tcl_dispatch.c#L411)
[- bypasses error handling](https://github.com/MDSplus/mdsplus/blob/ad08c9cc7c88f7eed946bf06f2bd6a90e2e1a25f/tcl/tcl_dispatch.c#L418)

Because the macros treat `SsINTERNAL` as success, the error bypasses several error handling sections in the call chain.   (The analysis of GetAnswerInfoTS is not shown above, but is similar to that for SendArg.)

**Expected behavior**
Any function that can return `SsINTERNAL`, `C_ERROR` or other non-MDSplus codes, should be followed by a status check that also includes those values.  In the above example, the following changes should be made.
- `if (STATUS_OK)` becomes `if ((status != SsINTERNAL) && STATUS_OK)`
- `if (STATUS_NOT_OK)` becomes `if (status == SsINTERNAL) || STATUS_NOT_OK)`

The above example is the fifth or so instance of this bug that I have seen in the MDSplus source code.   There are undoubtedly more bugs like this lurking in the source code.   However, they are likely in seldom exercised error handling code sections.   (Bugs like this that affected the mainstream workflow would have been reported by customers years ago and are surely already fixed.)

**Screenshots**
n/a

**Additional context**
There are three strategies for dealing with this issue.

1. **Don't Propagate** -- Don't allow non-MDSplus error codes to ripple up the entire call chain.  Limit it to just the first calling routine.  Find all instances of `SsINTERNAL` and `C_ERROR` in the source code.   Create a list of all the associated functions.   Then create a second list of all routines that call the functions on the first list.   Make sure that all functions on the second list correctly process `SsINTERNAL` and `C_ERROR` and only return MDSplus error codes.   This is the safest approach, but also the most time consuming to implement.
2. **Redefine Macros** -- Change the definition of the macros (IS_OK, IS_NOT_OK, etc) to also check for `SsINTERNAL` and `C_ERROR`.   A simple and easy change, but is riskier because it is a global change.
3. **Redefine SsINTERNAL** -- Redefine `SsINTERNAL` to be `-2` (i.e. `0xFFFFFFFE`) so that if it is inadvertently passed into the macros, it will be treated as a failure and not as a success.   (This is the same suggestion made in the comments on PR 2740, https://github.com/MDSplus/mdsplus/pull/2740#discussion_r1563452509 .)  Also simple and easy, but risky because it is a global change.  

The "Don't Propagate" strategy should be feasible because the number of functions is fairly small.

- `SsINTERNAL` occurs 36 times, but in just 13 files
- `C_OK` is 23 times, 14 files
- `C_ERROR` is 67 times, 14 files
- `B_TRUE` is 20 times, 12 files
- `B_FALSE` is 36 times, 13 files
- `FALSE` and `TRUE` are too numerous to inspect each occurrence

**Comment by mwinkel-dev** [comment]
If the goal of `SsINTERNAL` is to alert the user to an extremely abnormal condition that should never arise during normal operation of MDSplus, then another approach is to throw an exception immediately.   Instead of functions using `return SsINTERNAL`, replace those lines with a macro that displays an error message and then terminates the program.

The issue with the current code is that it masks errors by treating `SsINTERNAL` as a flavor of success.  The current code "swallows" the `SsINTERNAL` and hides it from the user.   The user is thus unaware that an abnormal condition occurred.

**Comment by zack-vii** [comment]
SsInternal is used to flag a situation that requires internal handling. i think it is used in remoteacces of treeshr to indicate a remote operation failed because of x is not supported so the code will try method y... something like that.

**Comment by zack-vii** [comment]
that said i agree the ok, error, true, and false should bbe consolidated tto follow one schema. but earlier attempts failed is it was either unclear if the methods are used by users (so we cannot chamge behaviour) and it was not always clear if 0 means success or false. we ended up replacing 0, 1 and -1 with the names you listed in order to help with that. however, noone dared to switch true/false to success/error or vice versa because we were afrait we would break code. today we have a better separation of public and private headers. so its worth checking if former concerns still hold

**Comment by mwinkel-dev** [comment]
Hi @zack-vii -- Thanks for the history.   

If you would prefer to discuss this via a Zoom meeting, let me know and I'll arrange to be available at whatever time is convenient for you.  (There are times when a complex topic is easier to handle with a discussion than with the written word.)

Reverting to the written word, I will try again to communicate the issue.

In my opinion, there is a difference between what we expect `SsINTERNAL` to do and what the code actually does.

There are three separate issues here.
- I concur that there are risks in consolidating the error schemas.   And I am hesitant to recommend such consolidation, given the potential impact on the customers, plus the current level of regression testing.
- Please review all of the links to the lines of code given in the bug report.   Which portions of that code do you recommend be changed?   **In this example, the lowest-level routine can generate `SsINTERNAL` and none of the routines in the entire call stack check or handle the `SsINTERNAL`**.   (More accurately stated, they do check the status but ignore the `SsINTERNAL` because they view it as "success".)  Shouldn't that instead be considered an error?
- Throughout the MDSplus source code, there are other instances similar to the above example.   What is the "best practice" for fixing those code sections?

**Comment by mwinkel-dev** [comment]
Note that the `SsINTERNAL` error code is defined in `mdsshr_messages.xml` which is then used by `gen-messages-exceptions.py` to create the `mdsshr_message.h` file.

```
<facility name="Ss" value="0">
<status name="INTERNAL" value="-1" severity="Internal" text="This status is meant for internal use only, you should never have seen this message."/>
```

The format of an MDSplus error code is <facility:16 bits><message#:13 bits><severity:3 bits>.

It can be argued that the XML definition of `SsINTERNAL` is fine, and that what was intended was `0000FFFE` as the definition of `SsINTERNAL` instead of the current `FFFFFFFF` (aka `-1`).   If so, that would mean that `get-messages-exceptions.py` is not properly handling the `value=-1` field of the XML file.   

Also, `get-messages-exceptions.py` would have to redefine the `internal` serverity as `6` (a flavor of failure) instead of the existing `7` (a flavor of success).
[change this](https://github.com/MDSplus/mdsplus/blob/16549d875c108bf9f53ce116b9665ee0f495345b/deploy/gen-messages-exceptions.py#L321)

Regardless, many sites are using versions of MDSplus that define `SsINTERNAL` to be `-1`, thus cross-version testing must be done to ensure that the eventual fix is robust for existing customers.

**Comment by zack-vii** [comment]
the intend of SsInternal is it will never surface. it does not have a regular code as it does not specify an error, it is used to mark an internal state that triggers a required cleanup or other action that must be handled.
we used -1 as this was the original value of must-be-handled value. it is a typical error code in c functions. in order to mark spots were the intend was to mark the state for extra actions we introduced SsInternal, for those were it is a result of a c-like method we used C_ERROR. it was important to us to not change the values as were wanted to make sure we do not change the behavior yet. handling those in IS_OK is not advisable. IsInternal is specificly not an error so it may trigger the good branch until handled. handling must happen or its a bug. if you experience IsSuccess at a userlevel it is usually due to a device driver that returned -1 in terms of a C_ERROR.

In short sometimes status does not get a status. what would help here is defining status aa strict type, e.g. an enum with only allowed values defined by the xml. none status values would then be marked by the compiler. you still have a problem on the interfaces though, namely tdi, python and co.
